### ISSUE TITLE
ERROR: rsync_cleanup_after_native_cp_al() only works on directories

### ISSUE THREAD

**Issue by olalonde** [rsnapshot/rsnapshot#133]
Getting a weird error:

```
$ /usr/local/bin/rsnapshot -c ./rsnapshot.conf sync
$ /usr/local/bin/rsnapshot -c ./rsnapshot.conf alpha
----------------------------------------------------------------------------
rsnapshot encountered an error! The program was invoked with these options:
/usr/local/bin/rsnapshot -c ./rsnapshot.conf alpha
----------------------------------------------------------------------------
ERROR: rsync_cleanup_after_native_cp_al() only works on directories
ERROR: Error! cp_al("/var/backups/data/snapshots/.sync", "/var/backups/data/snapshots/alpha.0")
```

Strangely, when I have a look at the alpha.0, everything seems correctly synced up.

I'm generating my rsnapshot.conf file with:

``` bash
#!/bin/bash

# generate template file
cat > ./rsnapshot.conf <<EOF
config_version  1.2
# All snapshots will be stored under this root directory.
snapshot_root   ${PWD}/data/snapshots/
#uncomment on LINUX only
#cmd_cp     /bin/cp
cmd_rm      /bin/rm
cmd_rsync   /usr/bin/rsync
cmd_logger  /usr/bin/logger
#cmd_preexec    send email?
#cmd_postexec   send email?
#########################################
#     BACKUP LEVELS / INTERVALS         #
# Must be unique and in ascending order #
# e.g. alpha, beta, gamma, etc.         #
#########################################
retain  alpha   7
retain  beta    4
retain  gamma   12

verbose     2
loglevel    3
#logfile    /var/log/rsnapshot

lockfile    ${PWD}/rsnapshot.pid

sync_first  1

###############################
### BACKUP POINTS / SCRIPTS ###
###############################

# LOCALHOST
backup  ${PWD}/data/latest/ localhost/
EOF

rsnapshot -c ./rsnapshot.conf "$@"
```

rsnapshot 1.4.2 on OS X

**Comment by axot** [comment]
I try to use `cmd_cp    /usr/local/bin/gcp` solved this issue :-|

**Comment by sam-at-github** [comment]
[rsync_cleanup_after_native_cp_al()](https://github.com/rsnapshot/rsnapshot/blob/master/rsnapshot-program.pl#L5149) looks to be using a strange way of checking that the src and dest are directories. Most likely the cause of the issue.

**Comment by bekarau** [comment]
In the 1.3 series, there was a piece of code that explicitly put the / on the end of the src and dest variables.

It appears that a recent update removed that code whilst not removing the requirement of looking for them in rsync_cleanup_after_native_cp_al.

Manually putting the /'s back in in a few places does get around the error using native_cp_al with an addition of an rsync command to do some --delete's.

**Comment by bekarau** [comment]
Huh, it appears that this section is still there, but ```cp_al``` is just called from other places where similar code is not in place.

In reality, ```cp_al``` (and probably ```display_cp_al```) probably need refactoring so that they are consistent and that the arguments are always correctly terminated (given that they should only take directories anyway).

**Comment by mytchel** [comment]
I know this is an old bug but I am getting the same problem. The fix above of removing the test does not work as the problem is not checking that the files are directories but that the `$src` given to rsync needs to have a trailing slash to work properly. Without it rsync is copying the `$src` into `$dest` as `$dest/${basename $src}

**Comment by sgpinkus** [comment]
Looked at this a while back, and made a PR. Can't remember why decided not to merge. To me the solution is to rm the first redundant check in this series of checks https://github.com/rsnapshot/rsnapshot/blob/02c3fb873a688f27866a83ae477d87d8ae1e0cef/rsnapshot-program.pl#L5148):

```
	# make sure this is directory to directory
	if (($src !~ m/\/$/o) or ($dest !~ m/\/$/o)) {
		print_err("rsync_cleanup_after_native_cp_al() only works on directories", 2);
		return (0);
	}

	# make sure we have a source directory
	if (!-d "$src") {
		print_err("rsync_cleanup_after_native_cp_al() needs a valid source directory as an argument", 2);
		return (0);
	}

	# make sure we have a destination directory
	if (!-d "$dest") {
		print_err("rsync_cleanup_after_native_cp_al() needs a valid destination directory as an argument",
			2);
		return (0);
	}
```

Or is Perl's `-d` not reliable?

**Comment by DrHyde** [comment]
It's reliable! And yes, that first check is obviously redundant.

**Comment by mytchel** [comment]
But the problem is not that they need to be directories. It is that `$src` needs to have a trailing slash for rsync.

**Comment by sgpinkus** [comment]
@DrHyde @mytchel Ah of course. 

> That code checks that there is a trailing /, but it might be better to simply force the presence of a trailing slash by appending it if it's missing, after checking that both the source and dest are directories.

Seems like the better approach to me too.

**Comment by sgpinkus** [comment]
@DrHyde @mytchel new PR for this #267.

**Comment by DrHyde** [comment]
Fixed in #267
